sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultServiceSpec.java

```diff
@@ -369,6 +379,38 @@ public ServiceSpec parse(byte[] bytes) throws ConfigStoreException {
         public static final Collection<Class<?>> getDefaultRegisteredSubtypes() {
             return defaultRegisteredSubtypes;
         }
+
+        /**
+         * Custom deserializer for goal states to accomodate transition from FINISHED to ONCE/FINISH.
+         */
+        public static class GoalStateDeserializer extends StdDeserializer<GoalState> {
+
+            public GoalStateDeserializer() {
+                this(null);
+            }
+
+            protected GoalStateDeserializer(Class<?> vc) {
+                super(vc);
+            }
+
+            @Override
+            public GoalState deserialize(
+                    JsonParser p, DeserializationContext ctxt) throws IOException, JsonParseException {
+                String value = ((TextNode) p.getCodec().readTree(p)).textValue();
+
+                if (value.equals("FINISHED") || value.equals("ONCE")) {
+                    return GoalState.ONCE;
+                } else if (value.equals("FINISH")) {
+                    return GoalState.FINISH;
+                } else if (value.equals("RUNNING")) {
+                    return GoalState.RUNNING;
+                } else if (value.equals("UNKNOWN")) {
```

Comment:
Would it make sense to do something like this, where any unrecognized value maps to UNKNOWN? I don't know one way or the other. Could help with downgrade compatibility later but may lead to other problems...:
```
} else {
  logger.warn(some warning here);
  return GoalState.UNKNOWN;
}
```

Reply:
+1